File: src/services/tts/google.js
const textToSpeech = require('@google-cloud/text-to-speech');
const client = new textToSpeech.TextToSpeechClient();
const Duplex = require('stream').Duplex;

module.exports = async (req, res, next) => {
    // Construct the request
    const rate = +(req.query.rate || 8000);
    const request = {
        input: {},
        // Select the language and SSML Voice Gender (optional)
        voice: {
            languageCode: req.query.language || 'en-US',
            ssmlGender: (req.query.gender || 'NEUTRAL').toUpperCase()
        },
        // Select the type of audio encoding
        audioConfig: {
            sampleRateHertz: rate,
            audioEncoding: getCodec(rate)
        },
    };

    if (req.query.audioEncoding) {
        request.audioConfig.audioEncoding = req.query.sampleRateHertz;
    }
    if (+req.query.sampleRateHertz) {
        request.audioConfig.sampleRateHertz = +req.query.sampleRateHertz;
    }
    if (+req.query.speakingRate) {
        request.audioConfig.speakingRate = +req.query.speakingRate;
    }
    if (+req.query.pitch) {
        request.audioConfig.pitch = +req.query.pitch;
    }
    if (+req.query.volumeGainDb) {
        request.audioConfig.volumeGainDb = +req.query.volumeGainDb;
    }
    if (req.query.effectsProfileId) {
        request.audioConfig.effectsProfileId = (req.query.effectsProfileId || '').split(',');
    }


    if (req.query.textType === 'ssml') {
        request.input.ssml = req.query.text
    } else {
        request.input.text = req.query.text;
    }

    // Performs the Text-to-Speech request
    try {
        const [response] = await client.synthesizeSpeech(request);
        res.setHeader("Content-Type", getContentType(rate));
        let stream = new Duplex();
        stream.push(response.audioContent);
        stream.push(null);
        res.status(200);
        stream.pipe(res);
    } catch (e) {
        return next(e);
    }
};

const getCodec = (rate = 8000) => {
    if (rate > 16000) {
        return "MP3"
    }
    return "LINEAR16"
};

function getContentType(rate = 8000) {
    if (rate > 16000) {
        return "audio/mpeg"
    }
    return "audio/ogg"
}

File: tts/google.go
package tts

import (
	"bytes"
	"context"
	"io"
	"io/ioutil"

	texttospeech "cloud.google.com/go/texttospeech/apiv1"
	"google.golang.org/api/option"
	texttospeechpb "google.golang.org/genproto/googleapis/cloud/texttospeech/v1"
)

func Google(params TTSParams) (io.ReadCloser, *string, error) {
	// Instantiates a client.
	ctx := context.Background()
	var err error
	var client *texttospeech.Client

	if params.KeyLocation != "" {
		client, err = texttospeech.NewClient(ctx, option.WithCredentialsFile(params.KeyLocation))
	} else {
		client, err = texttospeech.NewClient(ctx)
	}
	if err != nil {
		return nil, nil, err
	}

	// Perform the text-to-speech request on the text input with the selected
	// voice parameters and audio file type.
	req := texttospeechpb.SynthesizeSpeechRequest{
		// Build the voice request, select the language code ("en-US") and the SSML
		// voice gender ("neutral").
		Voice: &texttospeechpb.VoiceSelectionParams{
			LanguageCode: params.Language,
		},
		// Select the type of audio file you want returned.
		AudioConfig: &texttospeechpb.AudioConfig{
			AudioEncoding: texttospeechpb.AudioEncoding_LINEAR16,
			//SpeakingRate:     0,
			//Pitch:            0,
			//VolumeGainDb:     0,
			SampleRateHertz: 8000,
			//EffectsProfileId: nil,
		},
	}

	if params.SpeakingRate != 0 {
		req.AudioConfig.SpeakingRate = params.SpeakingRate
	}

	if params.Pitch != 0 {
		req.AudioConfig.Pitch = params.Pitch
	}

	if params.VolumeGainDb != 0 {
		req.AudioConfig.VolumeGainDb = params.VolumeGainDb
	}

	if params.EffectsProfileId != nil {
		req.AudioConfig.EffectsProfileId = params.EffectsProfileId
	}

	switch params.Voice {
	case "MALE":
		req.Voice.SsmlGender = texttospeechpb.SsmlVoiceGender_MALE
	case "FEMALE":
		req.Voice.SsmlGender = texttospeechpb.SsmlVoiceGender_FEMALE
	case "NEUTRAL":
		req.Voice.SsmlGender = texttospeechpb.SsmlVoiceGender_NEUTRAL
	default:
		req.Voice.Name = params.Voice
	}

	v := "audio/ogg"
	if params.Format == "mp3" {
		v = "audio/mp3"
		req.AudioConfig.SampleRateHertz = 22050
		req.AudioConfig.AudioEncoding = texttospeechpb.AudioEncoding_MP3
	}

	// Set the text input to be synthesized.
	if params.TextType == "ssml" {
		req.Input = &texttospeechpb.SynthesisInput{
			InputSource: &texttospeechpb.SynthesisInput_Ssml{Ssml: params.Text},
		}
	} else {
		req.Input = &texttospeechpb.SynthesisInput{
			InputSource: &texttospeechpb.SynthesisInput_Text{Text: params.Text},
		}
	}

	resp, err := client.SynthesizeSpeech(ctx, &req)
	if err != nil {
		return nil, nil, err
	}

	r := ioutil.NopCloser(bytes.NewReader(resp.GetAudioContent()))
	client.Close() // FIXME

	return r, &v, nil
}
